Framework/PythonInterface/mantid/utils/absorptioncorrutils.py

```diff
@@ -32,174 +33,245 @@ def _getBasename(filename):
     else:
         raise ValueError("Unrecognized absorption correction method '{}'".format(abs_method))
 
-    # step_1: check memory to see if the ws is already there
-    # -- check SHA1
-    # NOTE: The new attribute absSHA1 should be there.
-    if mtd.doesExist(wsn_as):
-        found_as = mtd[wsn_as].run()["absSHA1"].value == sha1
-    if mtd.doesExist(wsn_ac):
-        found_ac = mtd[wsn_ac].run()["absSHA1"].value == sha1
-
-    # step_2: try to load from cache_file provided
-    if (not found_as) or (not found_ac):
-        # disk -> memory
-        if os.path.exists(cache_file_name):
-            # LoadNexus(Filename=cache_file_name, OutputWorkspace=wsn_as)
-            # if mtd[wsn_as].id() == "WorkspaceGroup":
-            #     UnGroupWorkspace(InputWorkspace=wsn_as)
-            wsntmp = f"{wsn_as}_wsg"
-            Load(Filename=cache_file_name, OutputWorkspace=wsntmp)
-            wstype = mtd[wsntmp].id()
-            if wstype == "Workspace2D":
-                RenameWorkspace(InputWorkspace=wsntmp, OutputWorkspace=wsn_as)
-            elif wstype == "WorkspaceGroup":
-                # there should be exactly two workspaces inside, ungroup them will
-                # restore them with the orignal name (wsn_as, wsn_ac)
-                UnGroupWorkspace(InputWorkspace=wsntmp)
+    # step_1: check memory
+    if mtd.doesExist(abs_wksp_sample):
+        found_abs_wksp_sample = mtd[abs_wksp_sample].run()["absSHA1"].value == sha1
+    if mtd.doesExist(abs_wksp_container):
+        found_abs_wksp_container = mtd[abs_wksp_container].run()["absSHA1"].value == sha1
+
+    # step_2: load from disk if either is not found in memory
+    cache_file_name = ""
+    if (not found_abs_wksp_sample) or (not found_abs_wksp_container):
+        for candidate in cache_files :
+            if os.path.exists(candidate):
+                cache_file_name = candidate
+                wsntmp = f"tmpwsg"
+                Load(Filename=cache_file_name, OutputWorkspace=wsntmp)
+                wstype = mtd[wsntmp].id()
+                if wstype == "Workspace2D":
+                    RenameWorkspace(InputWorkspace=wsntmp, OutputWorkspace=abs_wksp_sample)
+                elif wstype == "WorkspaceGroup":
+                    UnGroupWorkspace(InputWorkspace=wsntmp)
+                else:
+                    raise ValueError(f"Unsupported cached workspace type: {wstype}")
+                break
+
+        if( cache_files and not cache_file_name ):
+            raise( RuntimeError("No valid cache files specified") )
+
+    # step_3: check memory again
+    if mtd.doesExist(abs_wksp_sample):
+        found_abs_wksp_sample = mtd[abs_wksp_sample].run()["absSHA1"].value == sha1
+    if mtd.doesExist(abs_wksp_container):
+        found_abs_wksp_container = mtd[abs_wksp_container].run()["absSHA1"].value == sha1
+
+    return found_abs_wksp_sample, found_abs_wksp_container, abs_wksp_sample, abs_wksp_container, cache_file_name
```

Comment:
Here is an edge case where where things might go wrong:
- user provided three dirs (dir1, dir2, dir3)
- no cache was found in dir1 (scan and save) , dir2 (scan only)
- due to some human error, a cache file in dir3(scan only) has the wrong filename, which trick the loader into loading it even if it is not the correct cache file
- the SHA check in step3 catch this issue and mark both `found_*` to false
- now we are returning the cache_file_name pointing at dir3, which will leads to a saving error when we try go save the newly generated cache later.

Reply:
fixed